File: rust/src/browser/browser_profiles.rs
//! Discovery of installed browser profiles that contain cookie databases.

use std::collections::{BTreeMap, BTreeSet};
use std::fs;
use std::path::{Path, PathBuf};

use anyhow::{anyhow, Result};
use serde::Deserialize;

/// Browsers whose on-disk cookie stores can be imported.
pub const SUPPORTED_COOKIE_BROWSERS: [&str; 5] = ["chrome", "edge", "brave", "chromium", "firefox"];

/// Metadata for a cookie-bearing installed browser profile.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct BrowserProfile {
    /// Normalized browser name.
    pub browser: String,
    /// On-disk profile name.
    pub name: String,
    /// Human-readable profile name, when the browser supplies one.
    pub display_name: String,
    /// Absolute or home-relative profile directory.
    pub path: PathBuf,
    /// Whether the browser marks this profile as the default/last used one.
    pub is_default: bool,
}

/// Options for [`list_browser_profiles`].
#[derive(Debug, Clone)]
pub struct BrowserProfileOptions {
    /// Restrict discovery to one browser. `None` scans all supported browsers.
    pub browser: Option<String>,
    /// Home directory used to resolve conventional profile locations.
    pub home_dir: PathBuf,
    /// Platform path convention (`linux`, `darwin`, or `win32`).
    pub platform: String,
}

impl Default for BrowserProfileOptions {
    fn default() -> Self {
        Self {
            browser: None,
            home_dir: dirs::home_dir().unwrap_or_else(|| PathBuf::from(".")),
            platform: current_platform().to_string(),
        }
    }
}

impl BrowserProfileOptions {
    /// Restrict discovery to one installed browser.
    pub fn browser(mut self, browser: impl Into<String>) -> Self {
        self.browser = Some(browser.into());
        self
    }

    /// Override the home directory used for discovery.
    pub fn home_dir(mut self, home_dir: impl Into<PathBuf>) -> Self {
        self.home_dir = home_dir.into();
        self
    }

    /// Override the platform convention, primarily for portable tooling/tests.
    pub fn platform(mut self, platform: impl AsRef<str>) -> Self {
        self.platform = normalize_platform(platform.as_ref()).to_string();
        self
    }
}

pub(crate) fn current_platform() -> &'static str {
    normalize_platform(std::env::consts::OS)
}

pub(crate) fn normalize_platform(platform: &str) -> &str {
    match platform {
        "macos" => "darwin",
        "windows" => "win32",
        other => other,
    }
}

pub(crate) fn normalize_cookie_browser(browser: &str) -> Result<&str> {
    let normalized = if browser == "msedge" { "edge" } else { browser };
    if SUPPORTED_COOKIE_BROWSERS.contains(&normalized) {
        Ok(normalized)
    } else {
        Err(anyhow!(
            "Unsupported browser: {browser}. Expected one of {}",
            SUPPORTED_COOKIE_BROWSERS.join(", ")
        ))
    }
}

pub(crate) fn browser_profile_root(
    browser: &str,
    platform: &str,
    home_dir: &Path,
) -> Result<PathBuf> {
    let browser = normalize_cookie_browser(browser)?;
    let local = std::env::var_os("LOCALAPPDATA")
        .map(PathBuf::from)
        .unwrap_or_else(|| home_dir.join("AppData/Local"));
    let roaming = std::env::var_os("APPDATA")
        .map(PathBuf::from)
        .unwrap_or_else(|| home_dir.join("AppData/Roaming"));
    let support = home_dir.join("Library/Application Support");
    let root = match (normalize_platform(platform), browser) {
        ("darwin", "chrome") => support.join("Google/Chrome"),
        ("darwin", "edge") => support.join("Microsoft Edge"),
        ("darwin", "brave") => support.join("BraveSoftware/Brave-Browser"),
        ("darwin", "chromium") => support.join("Chromium"),
        ("darwin", "firefox") => support.join("Firefox"),
        ("win32", "chrome") => local.join("Google/Chrome/User Data"),
        ("win32", "edge") => local.join("Microsoft/Edge/User Data"),
        ("win32", "brave") => local.join("BraveSoftware/Brave-Browser/User Data"),
        ("win32", "chromium") => local.join("Chromium/User Data"),
        ("win32", "firefox") => roaming.join("Mozilla/Firefox"),
        (_, "chrome") => home_dir.join(".config/google-chrome"),
        (_, "edge") => home_dir.join(".config/microsoft-edge"),
        (_, "brave") => home_dir.join(".config/BraveSoftware/Brave-Browser"),
        (_, "chromium") => home_dir.join(".config/chromium"),
        (_, "firefox") => home_dir.join(".mozilla/firefox"),
        _ => unreachable!(),
    };
    Ok(root)
}

pub(crate) fn find_cookie_database(browser: &str, profile_path: &Path) -> Option<PathBuf> {
    if browser == "firefox" {
        let candidate = profile_path.join("cookies.sqlite");
        return candidate.is_file().then_some(candidate);
    }
    [
        profile_path.join("Network/Cookies"),
        profile_path.join("Cookies"),
    ]
    .into_iter()
    .find(|candidate| candidate.is_file())
}

#[derive(Debug, Default, Deserialize)]
struct LocalState {
    #[serde(default)]
    profile: LocalStateProfile,
}

#[derive(Debug, Default, Deserialize)]
struct LocalStateProfile {
    last_used: Option<String>,
    #[serde(default)]
    info_cache: BTreeMap<String, LocalStateProfileInfo>,
}

#[derive(Debug, Default, Deserialize)]
struct LocalStateProfileInfo {
    name: Option<String>,
}

fn list_chromium_profiles(browser: &str, root: &Path) -> Vec<BrowserProfile> {
    if !root.is_dir() {
        return Vec::new();
    }
    let state = fs::read_to_string(root.join("Local State"))
        .ok()
        .and_then(|contents| serde_json::from_str::<LocalState>(&contents).ok())
        .unwrap_or_default();
    let mut names = state
        .profile
        .info_cache
        .keys()
        .cloned()
        .collect::<BTreeSet<_>>();
    if let Ok(entries) = fs::read_dir(root) {
        for name in entries
            .flatten()
            .filter_map(|entry| entry.file_name().into_string().ok())
            .filter(|name| name == "Default" || name.starts_with("Profile "))
        {
            names.insert(name);
        }
    }
    let default_name = state.profile.last_used.as_deref().unwrap_or("Default");
    let only_default = names.len() == 1 && names.contains("Default");
    let mut profiles = names
        .into_iter()
        .filter_map(|name| {
            let path = root.join(&name);
            find_cookie_database(browser, &path)?;
            let display_name = state
                .profile
                .info_cache
                .get(&name)
                .and_then(|info| info.name.clone())
                .unwrap_or_else(|| name.clone());
            Some(BrowserProfile {
                browser: browser.to_string(),
                is_default: name == default_name || (only_default && name == "Default"),
                name,
                display_name,
                path,
            })
        })
        .collect::<Vec<_>>();
    profiles.sort_by_key(|profile| (!profile.is_default, profile.name.clone()));
    profiles
}

fn parse_ini(contents: &str) -> Vec<BTreeMap<String, String>> {
    let mut sections = Vec::new();
    let mut current: Option<BTreeMap<String, String>> = None;
    for raw_line in contents.lines() {
        let line = raw_line.trim();
        if line.starts_with('[') && line.ends_with(']') {
            if let Some(section) = current.take() {
                sections.push(section);
            }
            let mut section = BTreeMap::new();
            section.insert("section".into(), line[1..line.len() - 1].into());
            current = Some(section);
        } else if let (Some(section), Some((key, value))) = (current.as_mut(), line.split_once('='))
        {
            section.insert(key.trim().into(), value.trim().into());
        }
    }
    if let Some(section) = current {
        sections.push(section);
    }
    sections
}

fn list_firefox_profiles(root: &Path) -> Vec<BrowserProfile> {
    let contents = match fs::read_to_string(root.join("profiles.ini")) {
        Ok(contents) => contents,
        Err(_) => return Vec::new(),
    };
    let mut profiles = parse_ini(&contents)
        .into_iter()
        .filter(|section| {
            section
                .get("section")
                .is_some_and(|name| name.starts_with("Profile"))
        })
        .filter_map(|section| {
            let configured = PathBuf::from(section.get("Path")?);
            let path = if section.get("IsRelative").map(String::as_str) == Some("0") {
                configured
            } else {
                root.join(configured)
            };
            find_cookie_database("firefox", &path)?;
            let name = section
                .get("Name")
                .cloned()
                .or_else(|| path.file_name()?.to_str().map(String::from))?;
            Some(BrowserProfile {
                browser: "firefox".into(),
                name: name.clone(),
                display_name: name,
                path,
                is_default: section.get("Default").map(String::as_str) == Some("1"),
            })
        })
        .collect::<Vec<_>>();
    profiles.sort_by_key(|profile| (!profile.is_default, profile.name.clone()));
    profiles
}

/// Discover cookie-bearing profiles for Chrome, Edge, Brave, Chromium, and Firefox.
pub fn list_browser_profiles(options: BrowserProfileOptions) -> Result<Vec<BrowserProfile>> {
    let browsers = match options.browser.as_deref() {
        Some(browser) => vec![normalize_cookie_browser(browser)?],
        None => SUPPORTED_COOKIE_BROWSERS.to_vec(),
    };
    let mut profiles = Vec::new();
    for browser in browsers {
        let root = browser_profile_root(browser, &options.platform, &options.home_dir)?;
        if browser == "firefox" {
            profiles.extend(list_firefox_profiles(&root));
        } else {
            profiles.extend(list_chromium_profiles(browser, &root));
        }
    }
    Ok(profiles)
}

pub(crate) fn resolve_browser_profile(
    browser: &str,
    requested_profile: Option<&str>,
    options: &BrowserProfileOptions,
) -> Result<BrowserProfile> {
    let profiles = list_browser_profiles(options.clone().browser(browser))?;
    let selected = requested_profile
        .and_then(|requested| {
            profiles.iter().find(|profile| {
                profile.name == requested
                    || profile.display_name == requested
                    || profile.path.file_name().and_then(|name| name.to_str()) == Some(requested)
            })
        })
        .or_else(|| profiles.iter().find(|profile| profile.is_default))
        .or_else(|| profiles.first());
    selected.cloned().ok_or_else(|| {
        let detail = requested_profile
            .map(|profile| format!(" profile \"{profile}\""))
            .unwrap_or_else(|| " profile".into());
        anyhow!("Could not find a cookie database for {browser}{detail}")
    })
}

File: js/src/browser/browser-profiles.js
import { access, readdir, readFile } from 'node:fs/promises';
import os from 'node:os';
import path from 'node:path';

export const SUPPORTED_COOKIE_BROWSERS = [
  'chrome',
  'edge',
  'brave',
  'chromium',
  'firefox',
];

function platformPath(platform) {
  return platform === 'win32' ? path.win32 : path;
}

export function normalizeCookieBrowser(browser) {
  const normalized = browser === 'msedge' ? 'edge' : browser;
  if (!SUPPORTED_COOKIE_BROWSERS.includes(normalized)) {
    throw new Error(
      `Unsupported browser: ${browser}. Expected one of ${SUPPORTED_COOKIE_BROWSERS.join(', ')}`
    );
  }
  return normalized;
}

export function browserProfileRoot(
  browser,
  {
    platform = process.platform,
    homeDir = os.homedir(),
    environment = process.env,
  } = {}
) {
  const normalizedBrowser = normalizeCookieBrowser(browser);
  const pathApi = platformPath(platform);
  if (platform === 'darwin') {
    const support = pathApi.join(homeDir, 'Library', 'Application Support');
    const roots = {
      brave: pathApi.join(support, 'BraveSoftware', 'Brave-Browser'),
      chrome: pathApi.join(support, 'Google', 'Chrome'),
      chromium: pathApi.join(support, 'Chromium'),
      edge: pathApi.join(support, 'Microsoft Edge'),
      firefox: pathApi.join(support, 'Firefox'),
    };
    return roots[normalizedBrowser];
  }
  if (platform === 'win32') {
    const localAppData =
      environment.LOCALAPPDATA ?? pathApi.join(homeDir, 'AppData', 'Local');
    const roamingAppData =
      environment.APPDATA ?? pathApi.join(homeDir, 'AppData', 'Roaming');
    const roots = {
      brave: pathApi.join(
        localAppData,
        'BraveSoftware',
        'Brave-Browser',
        'User Data'
      ),
      chrome: pathApi.join(localAppData, 'Google', 'Chrome', 'User Data'),
      chromium: pathApi.join(localAppData, 'Chromium', 'User Data'),
      edge: pathApi.join(localAppData, 'Microsoft', 'Edge', 'User Data'),
      firefox: pathApi.join(roamingAppData, 'Mozilla', 'Firefox'),
    };
    return roots[normalizedBrowser];
  }
  const roots = {
    brave: pathApi.join(homeDir, '.config', 'BraveSoftware', 'Brave-Browser'),
    chrome: pathApi.join(homeDir, '.config', 'google-chrome'),
    chromium: pathApi.join(homeDir, '.config', 'chromium'),
    edge: pathApi.join(homeDir, '.config', 'microsoft-edge'),
    firefox: pathApi.join(homeDir, '.mozilla', 'firefox'),
  };
  return roots[normalizedBrowser];
}

async function pathExists(filePath) {
  try {
    await access(filePath);
    return true;
  } catch {
    return false;
  }
}

async function readJson(filePath) {
  try {
    return JSON.parse(await readFile(filePath, 'utf8'));
  } catch {
    return {};
  }
}

function chromiumCookiePaths(profilePath, pathApi = path) {
  return [
    pathApi.join(profilePath, 'Network', 'Cookies'),
    pathApi.join(profilePath, 'Cookies'),
  ];
}

export async function findCookieDatabase(browser, profilePath, platform) {
  const pathApi = platformPath(platform ?? process.platform);
  if (normalizeCookieBrowser(browser) === 'firefox') {
    const candidate = pathApi.join(profilePath, 'cookies.sqlite');
    return (await pathExists(candidate)) ? candidate : null;
  }
  for (const candidate of chromiumCookiePaths(profilePath, pathApi)) {
    if (await pathExists(candidate)) {
      return candidate;
    }
  }
  return null;
}

async function listChromiumProfiles(browser, root, platform) {
  if (!(await pathExists(root))) {
    return [];
  }
  const pathApi = platformPath(platform);
  const localState = await readJson(pathApi.join(root, 'Local State'));
  const infoCache = localState.profile?.info_cache ?? {};
  const names = new Set(Object.keys(infoCache));
  try {
    const entries = await readdir(root, { withFileTypes: true });
    for (const entry of entries) {
      if (
        entry.isDirectory() &&
        (entry.name === 'Default' || entry.name.startsWith('Profile '))
      ) {
        names.add(entry.name);
      }
    }
  } catch {
    return [];
  }

  const profiles = [];
  for (const name of names) {
    const profilePath = pathApi.join(root, name);
    if (!(await findCookieDatabase(browser, profilePath, platform))) {
      continue;
    }
    profiles.push({
      browser,
      name,
      displayName: infoCache[name]?.name ?? name,
      path: profilePath,
      isDefault:
        name === (localState.profile?.last_used ?? 'Default') ||
        (names.size === 1 && name === 'Default'),
    });
  }
  return profiles.sort(
    (left, right) =>
      Number(right.isDefault) - Number(left.isDefault) ||
      left.name.localeCompare(right.name)
  );
}

function parseIni(text) {
  const sections = [];
  let current;
  for (const rawLine of text.split(/\r?\n/)) {
    const line = rawLine.trim();
    const sectionMatch = /^\[([^\]]+)]$/.exec(line);
    if (sectionMatch) {
      current = { section: sectionMatch[1] };
      sections.push(current);
      continue;
    }
    const separator = line.indexOf('=');
    if (current && separator > 0 && !line.startsWith(';')) {
      current[line.slice(0, separator)] = line.slice(separator + 1);
    }
  }
  return sections;
}

async function listFirefoxProfiles(root, platform) {
  if (!(await pathExists(root))) {
    return [];
  }
  const pathApi = platformPath(platform);
  let sections;
  try {
    sections = parseIni(
      await readFile(pathApi.join(root, 'profiles.ini'), 'utf8')
    );
  } catch {
    const profilesRoot = pathApi.join(root, 'Profiles');
    try {
      const entries = await readdir(profilesRoot, { withFileTypes: true });
      sections = entries
        .filter((entry) => entry.isDirectory())
        .map((entry) => ({
          section: 'Profile',
          Name: entry.name,
          Path: entry.name,
          ProfilesRoot: profilesRoot,
        }));
    } catch {
      return [];
    }
  }

  const profiles = [];
  for (const section of sections.filter(({ section }) =>
    section.startsWith('Profile')
  )) {
    if (!section.Path) {
      continue;
    }
    const relativeRoot = section.ProfilesRoot ?? root;
    const profilePath =
      section.IsRelative === '0'
        ? section.Path
        : pathApi.resolve(relativeRoot, section.Path);
    if (!(await findCookieDatabase('firefox', profilePath, platform))) {
      continue;
    }
    const displayName = section.Name ?? pathApi.basename(profilePath);
    profiles.push({
      browser: 'firefox',
      name: displayName,
      displayName,
      path: profilePath,
      isDefault: section.Default === '1',
    });
  }
  return profiles.sort(
    (left, right) =>
      Number(right.isDefault) - Number(left.isDefault) ||
      left.name.localeCompare(right.name)
  );
}

/** Discover cookie-bearing profiles from installed browsers. */
export async function listBrowserProfiles({
  browser,
  platform = process.platform,
  homeDir = os.homedir(),
  environment = process.env,
} = {}) {
  const browsers = browser
    ? [normalizeCookieBrowser(browser)]
    : SUPPORTED_COOKIE_BROWSERS;
  const profiles = [];
  for (const candidate of browsers) {
    const root = browserProfileRoot(candidate, {
      platform,
      homeDir,
      environment,
    });
    profiles.push(
      ...(candidate === 'firefox'
        ? await listFirefoxProfiles(root, platform)
        : await listChromiumProfiles(candidate, root, platform))
    );
  }
  return profiles;
}

export async function resolveBrowserProfile(options) {
  const browser = normalizeCookieBrowser(options.browser);
  const profiles = await listBrowserProfiles({ ...options, browser });
  const requested = options.profile;
  const selected = requested
    ? profiles.find(
        ({ name, displayName, path: profilePath }) =>
          requested === name ||
          requested === displayName ||
          requested === platformPath(options.platform).basename(profilePath)
      )
    : (profiles.find(({ isDefault }) => isDefault) ?? profiles[0]);
  if (!selected) {
    const detail = requested ? ` profile "${requested}"` : ' profile';
    throw new Error(`Could not find a cookie database for ${browser}${detail}`);
  }
  return selected;
}
